File: frontend/src/api/optimization.ts
import apiClient from './client'

export type DesignType = 'patch' | 'slot' | 'fractal' | 'custom'
export type Algorithm = 'ga' | 'pso' | 'hybrid'

export interface OptimizationConfig {
  project_id: number
  design_type: DesignType
  algorithm: Algorithm
  population_size?: number
  generations?: number
  constraints?: Record<string, any>
}

export interface OptimizationHistory {
  generation: number
  best_fitness: number
  avg_fitness: number
}

export interface OptimizationResult {
  run_id: number
  best_candidate: {
    params: Record<string, number>
    fitness: number
    freq_error_ghz?: number
    bandwidth_error_mhz?: number
    gain_estimate_dBi?: number
    return_loss_dB?: number
  }
  history: OptimizationHistory[]
  metrics: Record<string, any>
}

export interface OptimizationRun {
  id: number
  project_id: number
  algorithm: Algorithm
  population_size: number
  generations: number
  status: 'pending' | 'running' | 'completed' | 'failed'
  best_fitness?: number
  log?: {
    history?: OptimizationHistory[]
  }
  created_at: string
  updated_at?: string
}

export interface DesignCandidate {
  id: number
  optimization_run_id?: number
  geometry_params: Record<string, any>
  fitness: number
  metrics: Record<string, any>
  is_best: boolean
  created_at: string
}

export const optimizationApi = {
  start: async (config: OptimizationConfig): Promise<OptimizationRun> => {
    const response = await apiClient.post('/optimize/start', config)
    return response.data
  },

  getRun: async (runId: number): Promise<OptimizationRun> => {
    const response = await apiClient.get(`/optimize/run/${runId}`)
    return response.data
  },

  getCandidates: async (runId: number): Promise<DesignCandidate[]> => {
    const response = await apiClient.get(`/optimize/run/${runId}/candidates`)
    return response.data
  },
}

export const simulationApi = {
  upload: async (projectId: number, file: File, tool: 'hfss' | 'cst' = 'hfss') => {
    const formData = new FormData()
    formData.append('file', file)
    const response = await apiClient.post(`/sim/upload/${projectId}?simulation_tool=${tool}`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    })
    return response.data
  },

  getCandidates: async (projectId: number): Promise<DesignCandidate[]> => {
    const response = await apiClient.get(`/sim/candidates/${projectId}`)
    return response.data
  },
}

File: backend/schemas/optimization.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, Any
from models.optimization import OptimizationAlgorithm, OptimizationStatus
from models.geometry import DesignType


class OptimizationConfig(BaseModel):
    project_id: int
    design_type: DesignType
    algorithm: OptimizationAlgorithm
    population_size: int = 30
    generations: int = 40
    constraints: Optional[dict[str, Any]] = None


class OptimizationHistory(BaseModel):
    generation: int
    best_fitness: float
    avg_fitness: float


class OptimizationResult(BaseModel):
    run_id: int
    best_candidate: dict[str, Any]
    history: list[OptimizationHistory]
    metrics: dict[str, Any]


class OptimizationRunResponse(BaseModel):
    id: int
    project_id: int
    algorithm: OptimizationAlgorithm
    population_size: int
    generations: int
    status: OptimizationStatus
    best_fitness: Optional[float] = None
    log: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class DesignCandidateResponse(BaseModel):
    id: int
    optimization_run_id: Optional[int] = None
    geometry_params: dict[str, Any]
    fitness: float
    metrics: dict[str, Any]
    is_best: bool
    created_at: datetime
    
    class Config:
        from_attributes = True





